### Your Name Here: Carter Ithier
### Your TF Name: Joe
### Graduate or Undergraduate Credit: Graduate

## This is e29 PSET3
##### DUE: March 28 2018
##### Harvard University DCE CSCI-E29
##### Instructor: Nenad Svrzikapa
##### Staff: Joe Palin, Kaleigh Douglas, Lena Hajjar, Phil Lodine, Alan Xie
### Instructions:

PSET3 is designed to solidify your Pandas skills.  Before you get started please review and understand our Academic Integrity policy.  We understand that for some of your work you may have to use external sorces.  Please cite and be a good acedemic citizen. Remember that All homework assignments, quizzes and exams must be your independent work!

Undergraduate Credit students may solve any 4 problems.

Tips from staff:  

1. Don't wait until the last day, our Psets require time and learning
2. Piazza is there for you, we are a team and our goal is not to test you, we want to teach you! Please ask questions!
3. Program a bit every day.  You can't learn these concepts by reading 2 chapters and attempting to solve the problems. Read a little, try the code, change the code,break the code, understand, learn.

## Problem 1: 
### Created by: Lena
### Undergraduate 2.5 points Graduate 2 points

# More Fun at Wal-mart

## Problem by Lena

I know what exactly what you're thinking.<br>
"That Wal-mart problem set was just SO riveting."<br>
Well, you're in luck, because we're using the Wal-mart dataset again! <br>
We're going to make a few functions to find out information about Wal-marts and Supercenters in each state. 

### Information about the dataset (literally the exact same information as from PSET2)
You can find the data in the file walmart.csv (source: https://github.com/plotly/datasets) The file has over 2000 records of Wal-Marts and Supercenters that have opened in the US since 1962. The dataset has the following columns:
<ul>
<li><b>storenum:</b> The store number</li>
<li><b>OPENDATE:</b> The opening date of the store in MM/DD/YYYY format</li>
<li><b>date_super:</b> The date that the store was converted into a supercenter in MM/DD/YYYY format. There may be empty values if the store was not converted into a supercenter</li>
<li><b>conversion:</b> The number of conversions from Wal-Mart to Supercenter that the store has gone through. </li>
<li><b>STREETADDR:</b> The store's street address</li>
<li><b>STRCITY:</b> The store's city</li>
<li><b>STRSTATE:</b> The store's state</li>
<li><b>ZIPCODE:</b> The store's zipcode</li>
<li><b>type_store:</b> The type of store. Possible values are Wal-mart and SuperCenter</li>
<li><b>LAT:</b> The store's latitude</li>
<li><b>LON:</b> The store's longitude</li>
<li><b>MONTH:</b> The month the store opened</li>
<li><b>DAY:</b> The day the store opened</li>
<li><b>YEAR:</b> The year the store opened</li>
</ul>


### Part 1: Getting the data
You should have downloaded walmart.csv from the assignment (or even the last assignment). In the next cell, load the csv into a dataframe. 

In [53]:
import numpy as np
import pandas as pd
from pandas import pivot_table
### Part 2: Exploring the dataimport pandas as pd

# load walmart.csv into a dataframe. Hint: use the read_csv function

walmart = pd.read_csv("walmart.csv")

In PSET2, we did some exploration of the dataset. Print first few rows using the .head() function, just so we can get a refresher of the data looks like.

In [3]:
# print the first few rows HINT: Use the .head() function 

walmart.head()

storenum OPENDATE date_super  conversion              STREETADDR  \
0         1   7/1/62     3/1/97         1.0        2110 WEST WALNUT   
1         2   8/1/64     3/1/96         1.0        1417 HWY 62/65 N   
2         4   8/1/65     3/1/02         1.0       2901 HWY 412 EAST   
3         8  10/1/67     3/1/93         1.0   1621 NORTH BUSINESS 9   
4         7  10/1/67        NaN         NaN  3801 CAMP ROBINSON RD.   

             STRCITY STRSTATE  ZIPCODE   type_store        LAT       LON  \
0             Rogers       AR    72756  Supercenter  36.342235 -94.07141   
1           Harrison       AR    72601  Supercenter  36.236984 -93.09345   
2     Siloam Springs       AR    72761  Supercenter  36.179905 -94.50208   
3          Morrilton       AR    72110  Supercenter  35.156491 -92.75858   
4  North Little Rock       AR    72118     Wal-Mart  34.813269 -92.30229   

   MONTH  DAY  YEAR  
0      7    1  1962  
1      8    1  1964  
2      8    1  1965  
3     10    1  1967  
4     10    1  1967

### Part 2: Finding information by state
It's good to have information about store openings in each state. In the next cell, find all the stores that are in Arkansas. walmart.STRSTATE == "AR" should be somewhere in your solution

In [4]:
# your code here

ARStores = walmart.loc[walmart["STRSTATE"] == "AR"]

#Let's briefly look at your result
ARStores.head()

storenum OPENDATE date_super  conversion              STREETADDR  \
0         1   7/1/62     3/1/97         1.0        2110 WEST WALNUT   
1         2   8/1/64     3/1/96         1.0        1417 HWY 62/65 N   
2         4   8/1/65     3/1/02         1.0       2901 HWY 412 EAST   
3         8  10/1/67     3/1/93         1.0   1621 NORTH BUSINESS 9   
4         7  10/1/67        NaN         NaN  3801 CAMP ROBINSON RD.   

             STRCITY STRSTATE  ZIPCODE   type_store        LAT       LON  \
0             Rogers       AR    72756  Supercenter  36.342235 -94.07141   
1           Harrison       AR    72601  Supercenter  36.236984 -93.09345   
2     Siloam Springs       AR    72761  Supercenter  36.179905 -94.50208   
3          Morrilton       AR    72110  Supercenter  35.156491 -92.75858   
4  North Little Rock       AR    72118     Wal-Mart  34.813269 -92.30229   

   MONTH  DAY  YEAR  
0      7    1  1962  
1      8    1  1964  
2      8    1  1965  
3     10    1  1967  
4     10    1  1967

Awesome, now we have all the stores in Arkansas. Now, make two more dataframes: one with all of the Supercenters in Arkansas, and one with all of the Wal-marts in Arkansas. You should be using the "type_store" column.

In [5]:
# your code here

# find all supercenters
ARSuper = ARStores.loc[ARStores["type_store"] == "Supercenter"]

# find all wal-marts
ARWalmart = ARStores.loc[ARStores["type_store"] == "Wal-Mart"]


At this point, I'm truly just interested in the YEAR information, so make 2 series of just the YEAR Columns for both Supercenters and Walmarts

In [8]:
# Your code here

#Find all YEARS that Supercenters Opened
ARSuperYears = ARSuper["YEAR"]

#Find all YEARS that Walmarts opened
ARWalmartYears = ARWalmart["YEAR"]


Now that we have all the years, let's find out when the first store opened, when the last store opened, and how many stores there are in total. You should use the .min(), .max() and .count() functions! Hint: you'll have to compare the results of min and max for the two Series to find out when the first and last store in the whole state opened up. 

In [9]:
# your code here 

# find min of ARSuperYears
ARSuperMin = ARSuperYears.min()

# find min of ARWalmartYyears
ARWalmartMin = ARWalmartYears.min()

# find max of ARSuperYears
ARSuperMax = ARSuperYears.max()

# find max of ARWalmartYyears
ARWalmartMax = ARWalmartYears.max()

# find count of ARSuperYears
ARSuperCount = ARSuperYears.count()

# find count of ARWalmartYyears
ARWalmartCount = ARWalmartYears.count()

# find First store tha opened in AR
# https://stackoverflow.com/questions/3357369/maximum-of-2-numbers
ARFirst = min(ARSuperMin, ARWalmartMin)

# find Last store tha opened in AR
ARLast = max(ARSuperMax, ARWalmartMax)

# find total number of stores in AR
ARCount = ARSuperCount + ARWalmartCount

#print results
print("The first Supercenter opened in ", ARSuperMin)

print("The first Walmart opened in ", ARWalmartMin)

print("The last Supercenter opened in ", ARSuperMax)

print("The last Walmart opened in ", ARWalmartMax)

print("The first store opened in ", ARFirst)

print("The last store opened in ", ARLast)

print("The total number of Supercenters  is ", ARSuperCount)

print("The total number of Wal-marts  is ", ARWalmartCount)

print("The total number of stores is ", ARCount)


The first Supercenter opened in  1962
The first Walmart opened in  1967
The last Supercenter opened in  2005
The last Walmart opened in  1982
The first store opened in  1962
The last store opened in  2005
The total number of Supercenters  is  60
The total number of Wal-marts  is  21
The total number of stores is  81


Now that we know how to find all this information, write a function called StateStoreData that takes a state abbreviation as the argument, and it returns a dictionary of state, total_stores, total_supercenters, total_walmarts, first_store, and last_store. 

Note: depending on how you solve this problem, you may get a warning from python because of the indexing. You can suppress the warnings by running the cell below. 

In [10]:
import warnings
warnings.filterwarnings('ignore')

In [16]:
def StateStoreData(state):
    ARStores = walmart.loc[walmart["STRSTATE"] == state]
    
    # find all supercenters
    ARSuper = ARStores.loc[ARStores["type_store"] == "Supercenter"]

    # find all wal-marts
    ARWalmart = ARStores.loc[ARStores["type_store"] == "Wal-Mart"]
    
    #Find all YEARS that Supercenters Opened
    ARSuperYears = ARSuper["YEAR"]

    #Find all YEARS that Walmarts opened
    ARWalmartYears = ARWalmart["YEAR"]

    # your code here
    # find min of ARSuperYears
    ARSuperMin = ARSuperYears.min()

    # find min of ARWalmartYyears
    ARWalmartMin = ARWalmartYears.min()

    # find max of ARSuperYears
    ARSuperMax = ARSuperYears.max()

    # find max of ARWalmartYyears
    ARWalmartMax = ARWalmartYears.max()

    # find count of ARSuperYears
    ARSuperCount = ARSuperYears.count()

    # find count of ARWalmartYyears
    ARWalmartCount = ARWalmartYears.count()

    # find First store tha opened in state
    # https://stackoverflow.com/questions/3357369/maximum-of-2-numbers
    ARFirst = min(ARSuperMin, ARWalmartMin)

    # find Last store tha opened in state
    ARLast = max(ARSuperMax, ARWalmartMax)

    # find total number of stores in state
    ARCount = ARSuperCount + ARWalmartCount
    
    dictionary = {"state": state}
    dictionary["total_stores"] = ARCount
    dictionary["total_supercenters"] = ARSuperCount
    dictionary["total_walmarts"] = ARWalmartCount
    dictionary["first_store"] = ARFirst
    dictionary["last_store"] = ARLast
    return dictionary
    
    

Pass "AR" into the funciton you just made to make sure you get the results as you did before you made the function.

In [17]:
# your code here
StateStoreData("AR")

{'first_store': 1962,
 'last_store': 2005,
 'state': 'AR',
 'total_stores': 81,
 'total_supercenters': 60,
 'total_walmarts': 21}

Despite the fact that there are at least 3 Wal-marts within 20 miles of my home in MA, this dataset doesn't have any information about MA stores. Does this function fail if you try to pass "MA"? <br>
<b>Answer HERE:</b> No it does not fail. It puts NAN for first_store and last_store, "MA" for state, and 0 for the other fields

In [19]:
# your code here
StateStoreData("MA")


{'first_store': nan,
 'last_store': nan,
 'state': 'MA',
 'total_stores': 0,
 'total_supercenters': 0,
 'total_walmarts': 0}

### Part 3: Using a pivot table
You can also solve this problem by using a pivot table. Make a pivot table with STRSTATE as an index, type_store as the columns, YEAR as value, and have the aggfuncitons be min, max, and count. <br>
<b>Note</b>: There is no np.count funciton, but there is an np.count_nonzero function

In [20]:
# your code here
walmartpivot= walmart.pivot_table(values = "YEAR", index = "STRSTATE", columns = "type_store", aggfunc = [min, max, np.count_nonzero])
walmartpivot

min                  max          count_nonzero         
type_store Supercenter Wal-Mart Supercenter Wal-Mart   Supercenter Wal-Mart
STRSTATE                                                                   
AL                1979     1981        2006     1991            76       14
AR                1962     1967        2005     1982            60       21
AZ                1988     1988        2005     1997            40       15
CA                1990     1990        2006     2006            13      146
CO                1985     1986        2006     2000            43       13
DE                1991     1991        2003     2006             4        4
FL                1982     1983        2006     2000           127       48
GA                1981     1981        2006     1998            99       15
IA                1983     1984        2006     1992            37       18
ID                1992     1992        2005     1993            14        3
IL                1977     1977        2006     2006            51       79
IN                1983     1985        2005     2000            61       27
KS                1971     1970        2003     1996            35       18
KY                1974     1978        2005     1993            59       22
LA                1970     1975        2005     1989            60       23
MD                1991     1991        2001     2005             8       33
MI                1990     1989        2006     2005            40       37
MN                1986     1987        2005     2005            21       31
MO                1968     1971        2006     2004            77       40
MS                1974     1975        2005     1989            53       12
MT                1992     1994        2001     1998             7        4
NC                1983     1987        2006     1997            78       34
ND                1990     1990        1990     1992             1        7
NE                1984     1982        2005     1991            23        3
NM                1983     1985        2005     1986            26        3
NV                1991     1990        2006     1998            15        8
NY                1991     1991        2006     2005            35       48
OH                1988     1989        2006     2004            72       52
OK                1968     1975        2003     1990            53       29
OR                1992     1992        2005     2005            12       17
PA                1990     1991        2006     2006            70       46
SC                1981     1981        2006     1995            49       14
SD                1990     1990        2006     1990            10        1
TN                1974     1973        2006     1995            81       16
TX                1975     1976        2006     1999           253       62
UT                1990     1992        2004     1995            26        2
VA                1987     1984        2006     2004            61       20
WA                1993     1994        2006     2005            19       20
WI                1986     1985        2005     2005            43       34
WV                1989     1990        2006     1994            26        6
WY                1989     1991        1992     1991             8        1

Now, return the row with the Arkansas data.  You should be looking at the index. 

In [25]:
#Your Code here
# https://stackoverflow.com/questions/11973741/filtering-and-selecting-from-pivot-tables-made-with-python-pandas

ARPivot = walmartpivot.ix["AR"]
ARPivot

               type_store 
min            Supercenter    1962
               Wal-Mart       1967
max            Supercenter    2005
               Wal-Mart       1982
count_nonzero  Supercenter      60
               Wal-Mart         21
Name: AR, dtype: int64

You can use iloc to return values of specific cells. To test this, return the value of the count of Wal-Marts.

In [28]:
ARWalmartCount = ARPivot.iloc[5]
print(ARWalmartCount)

21


Now make a new function called StateStoreData2 that takes a state abbreviation as the argument, and it returns a dictionary of state, total_stores, total_supercenters, total_walmarts, first_store, and last_store. You must use the pivot table and iloc!


In [29]:
def StateStoreData2(state):
    # your code here
    walmartpivot= walmart.pivot_table(values = "YEAR", index = "STRSTATE", columns = "type_store", aggfunc = [min, max, np.count_nonzero])
    ARPivot = walmartpivot.ix[state]
    total_supercenters = ARPivot.iloc[4]
    total_walmarts = ARPivot.iloc[5]
    total_stores = total_supercenters + total_walmarts
    first_store = min(ARPivot.iloc[0], ARPivot.iloc[1])
    last_store = max(ARPivot.iloc[2], ARPivot.iloc[3])
    return {"state": state, "total_stores": total_stores, "total_supercenters": total_supercenters, "total_walmarts": total_walmarts, "first_store": first_store, "last_store": last_store}

In [30]:
StateStoreData2("AR")

{'first_store': 1962,
 'last_store': 2005,
 'state': 'AR',
 'total_stores': 81,
 'total_supercenters': 60,
 'total_walmarts': 21}

## Problem 2: 
### Created by: Phil
### Undergraduate 2.5 points Graduate 2 points

# Exploring Voting Results by Method
## PSet3 Problem by Phil

As some of us may remember, the 2000 Presidential election was extremely close and hotly contested. Much of the drama occurred in the state of Florida, where many ballots had to be recounted by hand. As the recount progressed, the nation learned of some of the shortcomings of various schemes of ballot and voting machine design. Some of the factors that were investigated included the type of mechanism used to count votes, and the design of the ballot itself.

In the Florida election, each county used one of five methods to count votes:
* Optical scanning of a paper ballot marked by filling in an oval next to the candidate's name
* Votamatic machines, which require the voter to use a stylus to press out a perforated rectangle (a "chad") in a standard IBM punch card
* Datavote machines, which use a similar mechanism to that of the Votamatic
* Hand counting of a ballot filled out manually by the voter
* Lever machines, which mechanically count the positions on a set of levers flipped by the voter

Hand counting and lever machines are older methods that are rarely used today. Recent trends have been away from punch-card ballots toward either optical scanning or direct entry on an electronic touch-screen. In 2000, Florida used almost entirely either optical or punch-card systems.

One significant problem in counting votes occurs when a voter fills out a ballot incorrectly. In many local elections one can vote for two or more candidates for a given position, up to a maximum. (For example, you might see an instruction like "Vote for any THREE" next to a list of ten candidates for three open positions on the local school board. If a voter's ballot shows choices for only one or two candidates in a multiple-choice race, or shows no choice at all for a single-choice race like President, this may have been his or her intention; *OR* the ballot may have been incorrectly marked, or may have been counted incorrectly by a machine. This situation is called an "undervote". Similarly, the voter's ballot may show more than the allowed number of candidates, an "overvote".

Another factor that can cause problems is whether the candidates' names are listed in a single column or in two columns. The two-column arrangement (the infamous "butterfly ballot") arranged the names to the left and right of a single column of hole-punch locations, accommodating the small width of the IBM punch card placed below the sheet that lists the candidates. This arrangement may make it more difficult to identify the correct location to punch, leading to ballots whose markings do not reflect the voter's intention. 

While undervotes and overvotes may not invalidate an entire ballot, how often they occur can tell us something about the relative reliability of different technologies and ballot designs.

Examples of the various ballots used recently in the U.S. can be seen here: https://www.verifiedvoting.org/resources/voting-equipment/. Images of the one- and two-column ballot shapes can be seen here: www.nytimes.com/images/2001/11/12/politics/recount/index_BALLOT.html

Let's look at some of the results compiled during the Florida 2000 recount. Note that there were several minor-party candidates besides Bush and Gore, but we'll focus on those two.
The fl2000.csv file contains the vote counts for each candidate in each county, along with:
* the type of technology used to count votes
* the number of columns on the ballot
* the number of undervotes and overvotes

This data was compiled by Brett Presnell of the University of Florida Statistics Department.

Let's load the file (available on Canvas) the usual way and see what it looks like:

In [31]:
import pandas as pd
import numpy as np
df = pd.read_csv('fl2000.csv')

In [32]:
df.head()

county technology columns under over   Bush   Gore Browne Nader Harris  \
0    string     string     int   int  int    int    int    int   int    int   
1   Alachua    Optical       1   217  105  34124  47365    658  3226      6   
2     Baker    Optical       1    79   46   5610   2392     17    53      0   
3       Bay    Optical       1   541  141  38637  18850    171   828      5   
4  Bradford    Optical       2    41  695   5414   3075     28    84      0   

  Hagelin Buchanan McReynolds Phillips Moorehead Chote McCarthy  
0     int      int        int      int       int   int      int  
1      42      263          4       20        21     0        0  
2       3       73          0        3         3     0        0  
3      18      248          3       18        27     0        0  
4       2       65          0        2         3     0        0

Hmmm... that first row looks odd. It looks like the data includes a helpful line at the top giving the data type of each column. This might be helpful for some users, but we have Pandas, which can figure this out itself. Not only that, but the presence of the strings "int" in every column but the first two cause Pandas to refuse to parse the actual integers as integers. To see this, we can list the data types of each column:

In [33]:
df.dtypes

county        object
technology    object
columns       object
under         object
over          object
Bush          object
Gore          object
Browne        object
Nader         object
Harris        object
Hagelin       object
Buchanan      object
McReynolds    object
Phillips      object
Moorehead     object
Chote         object
McCarthy      object
dtype: object

Yep, this extra row is a problem. Look up the documentation for the read_csv() function and find out how to use the skiprows parameter to skip just this one row. (Hint: this parameter behaves differently depending on whether you supply a single integer or a list.) Once you've reloaded the dataset, show the first 5 lines again to verify your results.

In [43]:
# Your code here
# https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html
#https://stackoverflow.com/questions/27325652/python-pandas-read-csv-skip-rows-but-keep-header
df = pd.read_csv('fl2000.csv', skiprows = range(1,2))
df.head()

county technology  columns  under  over    Bush   Gore  Browne  Nader  \
0   Alachua    Optical        1    217   105   34124  47365     658   3226   
1     Baker    Optical        1     79    46    5610   2392      17     53   
2       Bay    Optical        1    541   141   38637  18850     171    828   
3  Bradford    Optical        2     41   695    5414   3075      28     84   
4   Brevard    Optical        1    277   136  115185  97318     643   4470   

   Harris  Hagelin  Buchanan  McReynolds  Phillips  Moorehead  Chote  McCarthy  
0       6       42       263           4        20         21      0         0  
1       0        3        73           0         3          3      0         0  
2       5       18       248           3        18         27      0         0  
3       0        2        65           0         2          3      0         0  
4      11       39       570          11        72         76      0         0

Use the dtypes method again to verify that the values are being interpreted correctly.

In [44]:
# Your code here
df.dtypes

county        object
technology    object
columns        int64
under          int64
over           int64
Bush           int64
Gore           int64
Browne         int64
Nader          int64
Harris         int64
Hagelin        int64
Buchanan       int64
McReynolds     int64
Phillips       int64
Moorehead      int64
Chote          int64
McCarthy       int64
dtype: object

The 'technology' column of your DataFrame is a pandas Series that tells the machine type or other method used by each county. Use the value_counts() method of the Series object to show how many counties use each of the five technologies.

In [45]:
# Your code here
df["technology"].value_counts()

Optical      41
Votomatic    15
Datavote      9
Hand          1
Lever         1
Name: technology, dtype: int64

Looking at the first few rows again, we see many columns of very low counts that we don't want to have cluttering up the display.

In [46]:
df.head()

county technology  columns  under  over    Bush   Gore  Browne  Nader  \
0   Alachua    Optical        1    217   105   34124  47365     658   3226   
1     Baker    Optical        1     79    46    5610   2392      17     53   
2       Bay    Optical        1    541   141   38637  18850     171    828   
3  Bradford    Optical        2     41   695    5414   3075      28     84   
4   Brevard    Optical        1    277   136  115185  97318     643   4470   

   Harris  Hagelin  Buchanan  McReynolds  Phillips  Moorehead  Chote  McCarthy  
0       6       42       263           4        20         21      0         0  
1       0        3        73           0         3          3      0         0  
2       5       18       248           3        18         27      0         0  
3       0        2        65           0         2          3      0         0  
4      11       39       570          11        72         76      0         0

Let's get rid of these by creating a new DataFrame that contains only the Bush, Gore, Browne, and Nader counts, along with the first five columns, which don't refer to candidates. Use the DataFrame's copy() method so that we can modify the new DataFrame later without generating the SettingWithCopyWarning message you might have seen. (Hint: this can be done in a single line of code by making a copy of a DataFrame composed of the first 9 columns of the original DataFrame; there are probably several ways to do this.) Assign the new DataFrame to the same variable *df* and show its first few lines.

In [48]:
# Your code here
# https://stackoverflow.com/questions/34682828/pandas-extracting-specific-selected-columns-from-a-dataframe-to-new-dataframe
df = df[["county", "technology", "columns", "under", "over", "Bush", "Gore", "Browne", "Nader"]].copy()
df.head()

county technology  columns  under  over    Bush   Gore  Browne  Nader
0   Alachua    Optical        1    217   105   34124  47365     658   3226
1     Baker    Optical        1     79    46    5610   2392      17     53
2       Bay    Optical        1    541   141   38637  18850     171    828
3  Bradford    Optical        2     41   695    5414   3075      28     84
4   Brevard    Optical        1    277   136  115185  97318     643   4470

So that we can see the magnitude of the under- and overvotes compared to the total number of votes cast, let's add a column called **All** that contains the sums of the Bush, Gore, Browne, and Nader vote counts. (Hint: once you've identified the columns you want to sum, you can sum them horizontally using sum() method's *axis* parameter.) If you find yourself re-executing your code as you debug it, you may find it handier to use an explicit [start:end] column range rather than an open-ended [start:] range, since adding a column will change the number of columns in the DataFrame each time you execute the cell. Once you've added the new column, show the first few lines of the DataFrame again.


In [49]:
# Your code here
# https://stackoverflow.com/questions/25748683/pandas-sum-dataframe-rows-for-given-columns
df['All'] = df[["Bush", "Gore", "Browne", "Nader"]].sum(axis = 1)
df.head()

county technology  columns  under  over    Bush   Gore  Browne  Nader  \
0   Alachua    Optical        1    217   105   34124  47365     658   3226   
1     Baker    Optical        1     79    46    5610   2392      17     53   
2       Bay    Optical        1    541   141   38637  18850     171    828   
3  Bradford    Optical        2     41   695    5414   3075      28     84   
4   Brevard    Optical        1    277   136  115185  97318     643   4470   

      All  
0   85373  
1    8072  
2   58486  
3    8601  
4  217616

We're almost there! Let's add 2 columns called **pct_under** and **pct_over** that show the percentages of the under- and overcounts, relative to the 'All' column. (Hint: these two lines will each just involve taking a column of the DataFrame, multiplying by 100 and dividing by another column.) 

In [50]:
# Your code here
df['pct_over'] = (df["over"] / df["All"])*100
df['pct_under'] = (df["under"] / df["All"])*100
df.head()

county technology  columns  under  over    Bush   Gore  Browne  Nader  \
0   Alachua    Optical        1    217   105   34124  47365     658   3226   
1     Baker    Optical        1     79    46    5610   2392      17     53   
2       Bay    Optical        1    541   141   38637  18850     171    828   
3  Bradford    Optical        2     41   695    5414   3075      28     84   
4   Brevard    Optical        1    277   136  115185  97318     643   4470   

      All  pct_over  pct_under  
0   85373  0.122990   0.254179  
1    8072  0.569871   0.978692  
2   58486  0.241083   0.925008  
3    8601  8.080456   0.476689  
4  217616  0.062495   0.127288

Lastly, let's build a pivot table that shows the total vote counts (the 'All' column) and the percentages of under- and overvotes across the entire state, broken down by voting machine technology and then ballot format (columns). The goal here is to see if anything jumps out at us that might indicate whether certain technologies or ballot types were more problematic than others. (Hints: your pivot table will have two indexes and three values. You will need to use different aggregation functions for the individual columns, since one is a sum and the other two are percentages.)

In [52]:
# Your code here
df.pivot_table(values = ["All", "pct_over", "pct_under"], index = ["technology", "columns"], aggfunc = {"All": "sum", "pct_over": "mean", "pct_under": "mean"})

All  pct_over  pct_under
technology columns                              
Datavote   1          39798  6.565091   0.907531
           2          31428  6.396980   0.841580
Hand       2           3787  6.152627   0.660153
Lever      1          61817  0.090590   0.286329
Optical    1        2135311  0.637282   0.568467
           2         175151  7.477317   0.724399
Votomatic  1        2796511  1.356059   1.576198
           2         692662  6.384276   2.146362

What preliminary conclusions might you draw from the various percentage values? It may be significant that the 2-column ballot format appears to higher percentages of under- and overvotes when used with any technology. On the other hand, the Datavote technology appears to have an even higher percentage of under- and overvotes with a 1-column ballot format. (There's no right or wrong answer here. We're not going to attempt any more sophisticated statistical analysis here -- we've done enough work for one day!)

Note: an annotated version of this dataset, with some comments by its author, is available here: http://hci.stanford.edu/courses/cs448b/data/florida-voting/fl2000.txt

## Problem 3: 
### Created by: Kaleigh
### Undergraduate 2.5 points Graduate 2 points

# Olympic Games - Summer and Winter

In this problem, we are exploring the olympic games records from both the winter and summer olympics. (The winter.csv file is the same as the file we used in Problem Set # 2.)  
Source: https://www.kaggle.com/the-guardian/olympic-games/data


In [54]:
#Imports
import numpy as np
import pandas as pd

## 1) Load Datasets  

- Load the "winter.csv" file as a pandas dataframe.  
- Load the "summer.csv" file as a pandas dataframe.  
- Display the **first 5** rows of the winter dataframe.  
- Display the **first 5** rows of the summer dataframe.

In [56]:
# Your Code Here

winter = pd.read_csv("winter.csv")
summer = pd.read_csv("summer.csv")

winter.head()


Year      City     Sport Discipline               Athlete Country Gender  \
0  1924  Chamonix  Biathlon   Biathlon           BERTHET, G.     FRA    Men   
1  1924  Chamonix  Biathlon   Biathlon        MANDRILLON, C.     FRA    Men   
2  1924  Chamonix  Biathlon   Biathlon   MANDRILLON, Maurice     FRA    Men   
3  1924  Chamonix  Biathlon   Biathlon       VANDELLE, André     FRA    Men   
4  1924  Chamonix  Biathlon   Biathlon  AUFDENBLATTEN, Adolf     SUI    Men   

             Event   Medal  
0  Military Patrol  Bronze  
1  Military Patrol  Bronze  
2  Military Patrol  Bronze  
3  Military Patrol  Bronze  
4  Military Patrol    Gold

In [57]:
summer.head()

Year    City     Sport Discipline             Athlete Country Gender  \
0  1896  Athens  Aquatics   Swimming       HAJOS, Alfred     HUN    Men   
1  1896  Athens  Aquatics   Swimming    HERSCHMANN, Otto     AUT    Men   
2  1896  Athens  Aquatics   Swimming   DRIVAS, Dimitrios     GRE    Men   
3  1896  Athens  Aquatics   Swimming  MALOKINIS, Ioannis     GRE    Men   
4  1896  Athens  Aquatics   Swimming  CHASAPIS, Spiridon     GRE    Men   

                        Event   Medal  
0              100M Freestyle    Gold  
1              100M Freestyle  Silver  
2  100M Freestyle For Sailors  Bronze  
3  100M Freestyle For Sailors    Gold  
4  100M Freestyle For Sailors  Silver

## 2) Merge Data
### 2.1) Add "Season" Column To Each Dataframe  
- For your winter dataframe, add a column with the heading "Season" where every value is "Winter"
- For your summer dataframe, add a column with the heading "Season" where every value is "Summer"

In [59]:
# Your Code Here
winter["Season"] = "Winter"
summer["Season"] = "Summer"

### 2.2) Merge The Winter And Summer Dataframes

In [60]:
# Your Code Here

olympics = pd.merge(winter, summer, on = "Sport")
olympics.head()



Empty DataFrame
Columns: [Year, City, Sport, Discipline, Athlete, Country, Gender, Event, Medal, Season]
Index: []

## 3) Clean Up Data
### 3.1) Remove Rows with Null Values  
In PSET 2, we confirmed there are no null values in the winter.csv dataset, however there are a few null values from the summer.csv dataset.  
Drop all rows with null values in your dataframe from Part 2.2.

In [ ]:
# Your Code Here


### 3.2) Remove Team Sport Bias  
This dataset includes every medal awarded to every individual athlete.  In team events, each medal (gold, silver and bronze) are awarded to multiple athletes.  (For example, the winning ice hockey team would receive ~20 gold medals.) In order to focus on the number of events a country has medaled in, rather than the number of athletes that have received medals, we're going to remove all "duplicate" medals from the same event.  

We'll do this by first dropping the 'Athlete' column and then removing all duplicate rows.

In [ ]:
# Drop the 'Athlete' Column
# Your Code Here



# Drop Duplcate Rows
# Your Code Here



## 4) How Many Gold Medals Has Jamaica Won in Athletics?   
### 4.1) Create a pivot table using the 'Country' and 'Sport' columns as indices and Medal (Gold, Silver, Bronze) as the table's columns.   The table's values will be the medal count for that particular country/sport/medal color.

**Display the head of the pivot table.**  
Note: Use the dataframe from Part 3.  
Note: The order of the Gold, Silver, Bronze columns does not matter.  
Hint: use fill_value = 0.  

Hint: The first three rows of your pivot table should look something like the following:  

    |        | Medal     | Bronze | Gold | Silver |  
    |Country | Sport     |        |      |        |  
    |AFG     | Taekwondo |   2    |   0  |    0   |  


In [ ]:
# Your Code Here


### 4.2) Access the pivot table to find the total number of Gold Medals that Jamaica (country code 'JAM') has won in the 'Athletics' Sport.

In [ ]:
# Your Code Here


## 5) Medal Counts Per Olympic Season
### 5.1) Create a pivot table using the 'Country' column as the table's index and the Seasons (Summer and Winter) as the table's columns.  The table's values will be the medal count for that particular country/season.

**Display the head of the pivot table.**  
Note: Use the dataframe from Part 3.  
Hint: use fill_value = 0.  

Hint: The first three rows of your pivot table should look something like the following:  

    |Season  | Summer | Winter |
    |Country |        |        | 
    |AFG     | 2      |   0    | 

In [ ]:
# Your Code Here


### 5.2) Display a plot of the pivot table from part 5.1. The plot should be a stacked horizontal bar graph, with the medal count on the x axis and the country codes on the y axis.  (Don't worry about plot titles/labels).     

In [ ]:
# Your Code Here


## Problem 4: 
### Created by: Alan
### Undergraduate 2.5 points Graduate 2 points

# Developing a Single-Game Win Model for the NBA (Part 1)

### Problem by Alan

Any basketball fans out here? (Personally, I'm more of an NFL kind of guy. How 'bout them Cowboys!) Nonetheless, get ready for some NBA excitement!

This problem (and its follow-ups in future problem sets) will take us from 0 to 1 in building a rudimentary model that can predict the likelihood of a team winning any single game during the NBA regular season. 

This week, we'll primarily focus on data exploration through the use of pivot tables. We'll also reinforce some of the DataFrame slicing and indexing concepts that we learned about last week. 

#### Step 1: Scrape and clean the data

Luckily, I've already gone to the trouble of collecting this data for you. (If you're curious, I collected the data by creating a local fork of [this existing API](https://github.com/jaebradley/basketball_reference_web_scraper) which scrapes the website [Basketball Reference](https://www.basketball-reference.com). Then, I rewrote the API to work again - the code on GitHub no longer functions and also relies on more outdated methods of web scraping such as the `urllib2` and `lxml` libraries instead of `requests` and `BeautifulSoup`.)

#### Step 2: Explore the data

This is an important step for us to eventually build a predictive model. We need to understand what our data looks like and what additional processing (if any) we might need to do in order to generate features for our model. I've provided you with three files: `2012_2017_box_scores.csv`, `2012_2017_schedules.csv`, and `2012_2017_season_player_stats.csv`. As you can see, the data provided is for the NBA regular season for a 6-year period between 2012 and 2017. (If you're wondering why the cut-off was somewhat arbitrarily decided to be 2012, that's because the 2011-2012 season was cut short by a player lockout.)

Let's get started by reading in our data using `pd.read_csv()`. Note that you might end up with an extra column if you don't pass in a value for the optional parameter `index_col`. This is because these files were previously created by `pd.to_csv()`, which by default exports a DataFrame's index as a regular column. The `index_col` parameter will allow us to read that column as the index, and our DataFrame will look normal again. This is especially useful for datasets with irregular indices!

In [ ]:
import pandas as pd
import numpy as np

In [70]:
# YOUR CODE HERE
# https://stackoverflow.com/questions/12960574/pandas-read-csv-index-col-none-not-working
box_scores = pd.read_csv("2012_2017_box_scores.csv", index_col = 0)
schedules = pd.read_csv("2012_2017_schedules.csv", index_col = 0)
player_stats = pd.read_csv("2012_2017_season_player_stats.csv", index_col = 0)


In [73]:
print(box_scores.shape)
print(schedules.shape)
print(player_stats.shape)

(138548, 22)
(6569, 5)
(3261, 25)


In [74]:
box_scores.head()


assists  blocks        date  defensive_rebounds  field_goal_attempts  \
0        3       0  2012-10-30                   9                   16   
1        9       2  2012-10-30                  11                    7   
2        6       3  2012-10-30                   9                   19   
3        5       0  2012-10-30                   5                   15   
4        4       1  2012-10-30                   1                   22   

   field_goals first_name  free_throw_attempts  free_throws  is_home  \
0           10     LeBron                    5            4     True   
1            3   Anderson                    3            3     True   
2            8        Pau                    8            6     True   
3            6       Paul                    9            9    False   
4           10     Dwyane                   11            9     True   

     ...     opponent  personal_fouls points  seconds_played  steals  team  \
0    ...          BOS               2     26          1732.0       2   MIA   
1    ...          WAS               4      9          2242.0       0   CLE   
2    ...          DAL               2     23          2419.0       0   LAL   
3    ...          MIA               3     23          2434.0       2   BOS   
4    ...          BOS               3     29          2107.0       2   MIA   

   three_point_field_goal_attempts three_point_field_goals  total_rebounds  \
0                                4                       2              10   
1                                0                       0              23   
2                                1                       1              13   
3                                4                       2               5   
4                                0                       0               3   

   turnovers  
0          0  
1          1  
2          2  
3          0  
4          4  

[5 rows x 22 columns]

In [75]:
schedules.head()

home_team_name  home_team_score           start_time  \
0  Cleveland Cavaliers             94.0  2012-10-30 23:00:00   
1   Los Angeles Lakers             91.0  2012-10-30 23:30:00   
2           Miami Heat            120.0  2012-10-31 00:00:00   
3        Chicago Bulls             93.0  2012-10-31 23:00:00   
4      Detroit Pistons             96.0  2012-10-31 23:30:00   

   visiting_team_name  visiting_team_score  
0  Washington Wizards                 84.0  
1    Dallas Mavericks                 99.0  
2      Boston Celtics                107.0  
3    Sacramento Kings                 87.0  
4     Houston Rockets                105.0

In [76]:
player_stats.head()

age  assists  blocks  defensive_rebounds  field_goals_attempted  \
0   22       11      15                  47                     75   
1   26       36      27                 128                    168   
2   27      206      11                 210                    905   
3   26        1       0                   1                      4   
4   27      192      91                 495                   1318   

   field_goals_made first_name  free_throws_attempted  free_throws_made  \
0                42     Quincy                     38                31   
1                72       Jeff                    100                65   
2               397      Arron                    223               191   
3                 2       Josh                      0                 0   
4               638   LaMarcus                    348               282   

   games_played  ...   points position  steals  team  \
0            29  ...      116       PF      13   TOR   
1            52  ...      209       PF      18   CHA   
2            64  ...     1057       SF      40   ORL   
3             3  ...        5       PG       0   DAL   
4            74  ...     1560       PF      62   POR   

   three_point_field_goals_attempted  three_point_field_goals_made turnovers  \
0                                  2                             1        17   
1                                  2                             0        32   
2                                240                            72       138   
3                                  2                             1         0   
4                                 14                             2       143   

   two_point_field_goals_attempted two_point_field_goals_made  year  
0                               73                         41  2012  
1                              166                         72  2012  
2                              665                        325  2012  
3                                2                          1  2012  
4                             1304                        636  2012  

[5 rows x 25 columns]

Examine each DataFrame and note in the below Markdown cell some initial observations. These can be some thoughts about the columns, the shape of the data, or the relationship that each DataFrame has with the others.

----

**Box Scores:** The box scores shape is (138548, 22). It has stats on each player by first name such as defensive rebounds, field goal attempts, and blocks.

**Schedules:** The schedules shape is (6569, 5). It lists when games were played (day and time), which teams were in the game, and what the score was.

**Season Player Stats:** The season player stats shape is (3261, 25). It lists stats on things like player age, what team they are on, how many assists, blocks, rebounds, and more. 

----

The most important task before us is to try to identify key features that might be predictive for our model. But what exactly would our model look like? Here, it's helpful to identify what our X and Y variables are. 

One analogy we can draw is to movies. If we want to build a box office forecasting model for opening weekend gross, then it's pretty clear that our Y variable is a number like `opening_weekend_box_office` (in dollars). Our X variables then might be things like `budget` (in dollars), `opening_number_of_theaters`, `holiday_weekend` (Boolean), and `sequel` (Boolean). Note that a qualitative variable like `director` might not immediately be useful for our model, as opposed to quantitative variables derived from `director` such as `average_director_lifetime_gross`. When we actually train our model, each observation in our train/test set is a single movie.

In the Markdown cell below, identify the X and Y variables for a model that aims to predict whether a home team will win an NBA regular season basketball game. The list of X variables does not have to be exhaustive but try to think of at least 5. You can refer to the Box Scores DataFrame for inspiration. We will tackle this topic of feature generation in greater detail in the future, so don't worry too much about it right now.

_Note: This is an important exercise because practical ML problems will require that you devote extensive thought to feature generation at the data exploration and processing step. In fact, proprietary features and data are often considered the "secret sauce" of most ML companies because open-source tools such as scikit-learn and Keras (combined with cheap AWS computing) have enabled anyone to perform sophisticated predictive analysis that was once considered extremely expensive and complex._

----
**X:** Three point field goals (int), total rebounds (int), free throws (int), field goals (int), assists (int), blocks (int)

**Y:** Win (boolean)

----

After going through all of the data, it looks like the Schedules DataFrame has all of the information we need to determine which games occurred on which dates and which teams won, as well as their home/visitor affiliation. Let's ignore this for now. Similarly, although the Season Player Stats DataFrame will likely be useful to us in the future, because it's a table of aggregate statistics there's not much we can do in terms of exploration right now. Instead, let's dive into the Box Scores DataFrame.

In [77]:
box_scores.columns

Index(['assists', 'blocks', 'date', 'defensive_rebounds',
       'field_goal_attempts', 'field_goals', 'first_name',
       'free_throw_attempts', 'free_throws', 'is_home', 'last_name',
       'offensive_rebounds', 'opponent', 'personal_fouls', 'points',
       'seconds_played', 'steals', 'team', 'three_point_field_goal_attempts',
       'three_point_field_goals', 'total_rebounds', 'turnovers'],
      dtype='object')

Looking at the columns, we see that there are a number of key stats. However, these stats are only available on an individual game-by-game basis. For our feature generation step, we may need this information aggregated across multiple games (but not at the season level). For instance, if we had a feature called `player_field_goal_percentage_to_date`, we would need to calculate the field goal percentage for a single player such as LeBron James for all of the games up to a certain date in the season.

If the Cavaliers are playing the Golden State Warriors in December, then it might be useful to calculate LeBron's field goal percentage for the season up to the date of the upcoming game. (We can also compare this to his previous season stats, which we can pull from the Season Player Stats DataFrame.)

Let's go ahead and try to do that! First, create a new column in Box Scores called `field_goal_percentage`. While you're at it, also create a new column called `full_name` that consists of the `first_name` and `last_name` fields concatenated with a space. Finally, convert the `date` column from type `str` to type `pd.Timestamp`.

In [80]:
# YOUR CODE HERE
# https://stackoverflow.com/questions/17134716/convert-dataframe-column-type-from-string-to-datetime
box_scores["field_goal_percentage"] = box_scores["field_goals"] / box_scores["field_goal_attempts"]
box_scores["full_name"] = box_scores["first_name"] + " " + box_scores["last_name"]
box_scores["date"] = pd.to_datetime(box_scores["date"])

Now, let's create a function called `get_player_FGP_to_date(df, player_full_name, date)`. In this function, create a pivot table on a date-filtered slice of your original data with `index` as `full_name` and `values` as the new `field_goal_percentage` column you created. Make sure that your function returns a float!

In [85]:
def get_player_FGP_to_date(df, player_full_name, date):
    """
    player_full_name: str of format 'LeBron James'
    date: str of format '2018-02-28' <== Cast this to pd.Timestamp using pd.to_datetime()
    @return result: float between 0 and 1
    """
    # YOUR CODE HERE
    date = pd.to_datetime(date)
    player_stats = df.loc[df["full_name"] == player_full_name]
    player_stats_date = player_stats.loc[player_stats["date"] < date]
    player_pivot = player_stats_date.pivot_table(values = "field_goal_percentage", index = "full_name")
    return player_pivot.ix[player_full_name]

_As a sanity check, the below cell should yield 0.55075._

In [86]:
get_player_FGP_to_date(box_scores, 'LeBron James', '2014-12-25') 

field_goal_percentage    0.550751
Name: LeBron James, dtype: float64

Let's write another function `get_player_TPP_to_date(df, player_full_name, date)` that generates a similar pivot table but for `three_point_percentage`. As before, create this column and then write the function, ensuring that it returns a float.

In [87]:
# YOUR CODE HERE
box_scores["three_point_percentage"] = box_scores["three_point_field_goals"] / box_scores["three_point_field_goal_attempts"]

In [88]:
def get_player_TPP_to_date(df, player_full_name, date):
    """
    player_full_name: str of format 'LeBron James'
    date: str of format '2018-02-28' <== Cast this to pd.Timestamp using pd.to_datetime()
    @return result: float between 0 and 1
    """
    # YOUR CODE HERE
    date = pd.to_datetime(date)
    player_stats = df.loc[df["full_name"] == player_full_name]
    player_stats_date = player_stats.loc[player_stats["date"] < date]
    player_pivot = player_stats_date.pivot_table(values = "three_point_percentage", index = "full_name")
    return player_pivot.ix[player_full_name]

_As a sanity check, the below cell should yield 0.33232._

In [89]:
get_player_TPP_to_date(box_scores, 'LeBron James', '2015-12-25') 

three_point_percentage    0.332321
Name: LeBron James, dtype: float64

Before we dive fully into feature generation, take a moment to think about one of the trickiest problems that we've had to deal with. Time traveling!

For example, if we're trying to predict the box office gross of Inception, and one of our X variables is `average_director_lifetime_gross`, we can't use movies directed by Christopher Nolan that came out after Inception. We have to date-filter our DataFrame to only include movies before the release date!

Similarly, if we want to calculate a player's three-point percentage to date, we can't include box scores from games that haven't happened yet. 

When we perform feature generation for each observation (each historical game), we will need to use functions such as the ones above to calculate all of our derived X variables. As you might expect, this will require us to loop through our DataFrame of games using `iterrows` or `apply`. 

#### Step 3: Feature generation

Tune in next week! Take some time to think about the features that you would want to generate!

## Problem 5: 
### Created by: Joe
### Undergraduate 2.5 points Graduate 2 points

# Gambling strategies

Gambling strategies.  Let's use Pandas to go gambling.  Last time we did some data cleaning, but we didn't really get to using our data.  This time we'll alter a little of our data processing, and then actually use our data.

In [ ]:
# damn pandas
import pandas as pd
import numpy as np
import datetime

Reading in our data again:

Easy removal of na_values using the na_values parameter:  Our bad values all have the string "null".

In [ ]:
# Read in our data with more help from the read_csv parameters
df = pd.read_csv("AAPL.csv",na_values = ["null"])
df.dtypes

We're only concerned with the Dates and Closing values, so get rid of the extra data:

In [ ]:
# remove the unwanted columns
df = 

# what does our data look like
df.head()

We also want to load in the Tesla and Netflix data.  Read the Tesla and Netflix Data in to df2 and df3:

In [ ]:
# read in the tesla and netflix data, and pare down the columns
df2 = 
df3 = 

# Pare down the unwanted columns
df2 = 
df3 = 

After we load all of the datasets, merge them together keeping just the dates and the closing values.  Rename the closing values so that we can distinguish the columns (hint, see section notes from a few weeks ago):

In [ ]:
# merge the data together into one file, and change the column
# names appropriately.
df = 
df = 

# And change the column names to fit a reasonable mold
df

In [ ]:
# The result of loading, paring down, and renaming columns should
# look something like this:
df.head()

Previously, we converted our dates to numbers, but this week we'll convert dates to time objects so we can use time deltas on them.  What do our times look like?

In [ ]:
# Our "Date" looks like strings.
df.loc[1:4,"Date"]

Datetime objects will let us more easily calculate differences in dates (as well as do power date manipulations like select particular months, days of the week, times of day, et cetera), so we'll convert our first column into date time objects.  We previously learned how to use the apply command for Series and data frames.  Use the apply command to change the dates in our dataframe to datetime objects.  Save the resulting series over the existing Date column.

The func required for your apply command should use the datetime strptime function.  See the documentation for more on how to use strptime: https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior

(If you have trouble with the strptime function, there is another method that uses a pandas library command which will let you infer date-times without knowing how strptime( ) works.

In [ ]:
# convert your "Date" objects into datetime objects.
df.loc[:,"Date"] = 

In [ ]:
# This should read datetime64 for Date, and float for everything else.
df.dtypes

Now that we have datetimes instead of strings, we can do easier calculations on them.  Create a mask of all the dates that were within the 10 days prior to March 31st, 2017 and show the data:

In [ ]:
# Probably want to create a datetime object for March 31st 2017
date = datetime.datetime.strptime(   )

# create a logical column with the datetime object and a timedelta object
mask = 

In [ ]:
# This command should output closing values for our three
# stocks from March 22nd to March 30th.
df.loc[mask,:]

We have a function which finds the 25th and 75th percentiles as we did last time.

In [ ]:
# this function takes a date, an index of dates in the form of a series,
# a data frame with close data, the number of days to reference, and quantiles.
# it returns the quantiles for the stocks in the data frame over the given
# interval.  Only three need to be defined.  The data-frame shouldn't include
# the dates, just the closing prices.
def percents( date_time , date_series , data_frame , look_back = 365 , quantiles = [25,75]):
    
    # turn the look back integer into a timedelta
    look_back = datetime.timedelta(look_back)

    # create a mask of acceptable dates
    mask = (-look_back < date_series - date_time) & (date_series - date_time <= datetime.timedelta(0))
    
    # find the percentile on the data that fits the range.  Axis = 0 allows runs the
    # function over columns instead of rows.
    nums = np.percentile( data_frame[mask],q = quantiles,axis = 0)
    
    # return the percentile calculation for each column of the data-frame.
    return nums


Now write a function that adds the quantiles to your data-frame for each of the stocks in our portfolio.  We've done this before, but this time, you can't use a for loop to iterate over the rows of the data if you want full credit.  apply( ) and loc( ) are the name of the game here.

In [ ]:
# Your function should take in a data frame, upper and lower quantiles,
# and possibly the number of days to look back when calculating the
# quantiles, and output a data frame with the quantiles added in successive
# columns.  If column 0 is the dates, 1,2,3 are our stock closes, columns
# 4,5,6 should be the quantiles for stocks 1,2,3.

def add_percentiles(df, quantiles = [25,75], look_back = 365):
    
    data_frame = df.copy()
    
    # YOUR CODE HERE

    return data_frame

In [ ]:
# how does our result look?  Should get something like this.
df1 = add_percentiles(df)
df1.head()

Great, we have some data to play with.  We're going to assess a betting strategy wherein we start with 1000 dollars.  If a stock we are following reaches the 75th percentile of the previous year's data, we'll sell half of our holdings.  If it reaches the 25th percentile, we'll buy stock according to half the cash we have on hand.

In [ ]:
# This function takes three stocks and applies our trading strategy
# across the portfolio.  It assumes a 3% monetary inflation rate.
# the return value is the value of the portfolio at each time 
# interval along the way.  The function assumes you have seven
# columns in your data frame, and that the 0th is dates, 1-3
# are stock names, and 4-6 are corresponding stock quantiles
def values(process):
    
    # Assumed inflation rate
    r = .03
    buy = ["neither","neither","neither"]

    then = process.iloc[0,0]

    
    portfolio = {"cash":1000,1:0,2:0,3:0}
    values = [1000]

    # values gives the rows of the data frame.
    for row in process.values:
        
        # set the time
        now = row[0]
        
        # grab the three stock close prices
        one,two,three = row[1:4]
        
        # iterating over each pair of stocks:
        for j in range(1,4):

            # extract the current close,low/high for each stock in turn
            close = row[j]
            low,high = row[j+3]

            # if the close is below the lower quantile, buy if we can buy
            if close < low and buy[j-1] != "bought":

                # find days passed
                days = (now-then).days

                # update the current
                then = now

                # update cash according to bank interest
                cash = portfolio["cash"] * np.exp(r*days/365)

                # buy stock with half of cash
                portfolio[j] += cash/2 / close

                # update cash
                portfolio["cash"] = cash/2

                buy[j-1] = "bought"

            # sell if needed
            elif close > high and buy[j-1] != "sold":

                # find days passed
                days = (now-then).days

                # update the current
                then = now

                # update cash according to bank interest
                cash = portfolio["cash"] * np.exp(r*days/365)

                # sell half of stock
                portfolio[j] /= 2 

                # update cash
                portfolio["cash"] += portfolio[j] * close

                buy[j-1] = "sold"
        
        worth = portfolio["cash"] + portfolio[1]*one + portfolio[2] * two + portfolio[3]*three
        values.append(worth)
        
    return values 

In [ ]:
# Simulate stock buys over a period
value1 = values(df1)

How much were we able to grow our money?

In [ ]:
# what was the last entry in each of the portfolio values.  Did we make any money? 
value1[-1]

That's a reasonable growth rate, although admittedly, not as good as could have been done with these three stocks.

What you need to do now is pick different quantiles to try in our trading strategy.  We default to 25 and 75 as the percentiles, but you could pick extremes like 1 and 99, 49 and 51, or something assymetric like 45 and 99.

Each set of quantiles you generate should give a new set of historic quantiles.  Take your best three sets of values and plot all the different sets on one line graph with plotly:

In [ ]:
# Creating quantiles for our data with multiple different quantile settings
df2 = add_percentiles(df,quantiles = )
df3 = add_percentiles(df,quantiles = )

df2.head()
df3.head()

value2 = values(df2)
value3 = values(df3)

## Plotting

Use the plotly library to generate visualizations for your exploraton.
Generate a line graph with the data you've generated.  Use the plotly basic line.  Make sure to label your graph!

(Reference for [plotly line-graphs](https://plot.ly/python/line-charts/#line-plot-modes).)

In [ ]:
# First plot the graph of the [25,75] quantile portfolio results.
# You'll need to do a little work getting reasonable x axis values







Great.  One scatter plot.  Now plot again, but plot three different sets of quantiles, and label each accordingly.  We want to be able to see how our choices performed.

In [ ]:
# Like the above, but with three sets of data overlaid:



Now that we've tested our algorithm on a few stocks of my choosing, do the same for 3 stocks of your choosing.  Repeat the analysis, and leave a blurb about what you find.  Is it possible to use this algorithm to make money on stocks that have mostly level value or maybe even lose value?

Note, it's really easy to get stock data from yahoo finance.  I got the apple data from [here](https://finance.yahoo.com/quote/AAPL/history?p=AAPL).  For a different stock, enter the company symbol of your choice.  You want historical data, and once you set the time frame, click the download data button.

In [ ]:
# Your stocks and analysis here: